pkg/ottl/ottlfuncs/func_trim.go

```diff
@@ -0,0 +1,45 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type TrimArguments[K any] struct {
+	Target      ottl.StringGetter[K]
+	Replacement ottl.Optional[string]
+}
+
+func NewTrimFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Trim", &TrimArguments[K]{}, createTrimFunction[K])
+}
+
+func createTrimFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*TrimArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("TrimFactory args must be of type *TrimArguments[K]")
+	}
+
+	return trim(args.Target, args.Replacement), nil
+}
+
+func trim[K any](target ottl.StringGetter[K], replacement ottl.Optional[string]) ottl.ExprFunc[K] {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		replacementString := replacement.Get()
+		val, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+		if replacement.IsEmpty() || replacementString == "" {
```

Comment:
Above line 34, lets to a check for if `replacement == ""` and error if it is.  Calling `Trim(name, "")` implies a nop, but we'd be trimming `" "`.

Reply:
@TylerHelmuth isn't an empty replacement string expected since the argument is optional?

Reply:
I'm saying we shouldn't allow a user to set the optional parameter to be "".

Reply:
IsEmpty returns false if the user sets the optional parameter to any value I thought, even default values like "".

Reply:
Yes, `replacement.isEmpty()` returns false with a default value of `""` for the `replacementString`.

However, the replacementString is set to `""` under two conditions

- When the replacement isn't provided (uses default `""`)
- When the `replacementString` is explicitly set to `""`

If we return an error when the `replacementString` is set to `""`, we would then make it a mandatory parameter

I've pushed a commit such that when the replacementString is set to `""` is either of the above conditions, there is no effect on the original string (default behavior).